### DISCUSSION TITLE
Interrupt disabling

### DISCUSSION THREAD

**Discussion by orbitcoms** [Mair/esp32-course#116] [Course Qeustions]
Hi,
Regarding the interrupts. The video showed an example of using interrupt to handle a button press,

First, I assume the interrupt gets cleared automatically when it leaves the interrupt handler?

In my application I need to stop the code processing once an interrupt has occurred until processing is complete.
I was wondering whether it would be ok to disable the interrupt within the ISR to prevent it re-triggering and then in my code I can enable it again when ready for more.

Or would it be better to use a flag like this? Also, can I set semaphore ok in here?

bool int_blocked = false;

void IRAM_ATTR my_isr_handler(void * args)
{
  if(!int_blocked) xSemaphoreGive(got_int);
}

process code..... Do stuff
int_blocked = false;

**Comment by Mair** [comment]
@orbitcoms 
1. I would avoid using global variables in this case as they are in different regions of memory
2. you must use the "fromISR" version of the semaphores also due to the memory boundary. not doing this can cause unpredictable behavior
3. if you look at the "fromISR versions you have an option to fire the code immediately or let the interrupt complete (pxHigherPriorityTaskWoken ). If you call your code immediately and in the normal thread of your app you can disable the interrupt https://www.freertos.org/a00119.html
4. have a look at the [de-bouncing ](https://github.com/Mair/esp32-course/blob/master/_10_gpio/_10_4_simple_debounce/main/main.c)example where this is done

**Reply by orbitcoms** [reply]
Thanks Mair,

So, it would look like this (I'm not concerned with pin number because there is only 1 gpio interrupt used).

static void IRAM_ATTR gpio_isr_handler(void * args)
{
	int pinNumber = (int) args;
	xQueueSendFromISR(interrupt_queue ,&pinNumber,NULL);
}

void process_interrupt(void *params)
{
	int pinNumber;
	while (true)
	{
		if (xQueueReceive(interrupt_queue, &pinNumber, portMAX_DELAY))
		{
			gpio_intr_disable(INT_PIN);
		}
	}
}

**Comment by Mair** [comment]
Yip except, if you are not interested in the in number you could also use a semaohore using xSemaphoreGiveFromISR in your interrupt and xSemaphoreTake in your task.
